### ISSUE TITLE
emacs eats up all my RAM while processing imports

### ISSUE THREAD

**Issue by RalfJung** [ProofGeneral/PG#499]
I am having trouble using PG on a particular Coq development, where when I step over the `Require` commands at the top of the file it never completes, instead eating up all of my RAM until the machine starts thrashing. When I killed it the emacs process was at more than 80% of my 16GB RAM. Running `make` in the same project to build it all with `coqc` works fine.

To reproduce, clone https://github.com/mit-pdos/perennial, initialize submodules (`git submodule update --init --recursive`), and then open pretty much any file and step to the end of the buffer in PG. I tried `external/Goose/github_com/mit_pdos/perennial_examples/replicated_block.v` and `src/program_logic/crash_weakestpre.v`.

This is with latest PG master and emacs 1:26.3+1-2 on Debian testing.

**Comment by RalfJung** [comment]
> does this happen with the default Proof General customization? Do
you have any auto-compilation features enabled?

Good idea! When I remove the "customization" block, the bug goes away. I'll try to bisect which is the "bad" option.

> When I try (with auto vos compilation) I immediately see an error
about some library not being in the load path and nothing goes
crazy, but I probably have a completely different Proof General
customization.

Did you do the submodule init? Also, does it help to run `make` on the console first? Which Coq version are you using?

**Comment by RalfJung** [comment]
So the problem is this line:
```
'(coq-compile-before-require t)
```
When I compile this out, it all works. But of course, I quite like this feature...

Here are my other coq/proof-related settings:
```
 '(coq-compile-parallel-in-background t)
 '(coq-compile-quick (quote no-quick))
 '(coq-one-command-per-line nil)
 '(proof-follow-mode (quote followdown))
 '(proof-next-command-insert-space nil)
 '(proof-splash-enable nil)
 '(proof-three-window-enable t)
```

**Comment by hendriktews** [comment]
I am sorry, I cannot reproduce. I did
```bash
git clone  https://github.com/mit-pdos/perennial
cd perennial/
git submodule update --init --recursive
make -j 4 src/program_logic/crash_weakestpre.vo
emacs -q -l ~/src/pg/499.el src/program_logic/crash_weakestpre.v
```
with 499.el containing
```emacs-lisp
(setq coq-compile-parallel-in-background t
      coq-compile-quick (quote no-quick)
      coq-one-command-per-line nil
      proof-follow-mode (quote followdown)
      proof-next-command-insert-space nil
      proof-splash-enable nil
      proof-three-window-enable t)

(load-file (expand-file-name "~/src/pg/master/generic/proof-site.el"))
```
In `src/program_logic/crash_weakestpre.v` I go to the bottom and
do C-c C-return. It takes a while, but eventually asserts
everything. Nothing goes crazy.

At `~/src/pg/master` I am at `5c3ebac1 New hook for early
prompt/output analyzis.` I use Coq 8.11.1.

Please provide more detailed instructions, preferably for a file
for which compilation takes less time.

And please do answer all the questions, in particular the one
about the `*Messages*` buffer.

**Comment by RalfJung** [comment]
> a file for which compilation takes less time

As mentioned before, `external/Goose/github_com/mit_pdos/perennial_examples/replicated_block.v` also shows this behavior. My guess is that all files do.

I am using `coq.8.12.dev` from opam. I'll see if using a different Coq version makes a difference. I'll also try to minimize my `.emacs` file so that I can share it.

I tried getting the `*Messages*` buffer, but unfortunately it is impossible: what I do is I open both that buffer and the Coq file side-by-side, and then C-c C-Return in the Coq file. But then it changes the window layout to the PG one (with the 3 buffers for file, goal, Coq output), hiding `*Messages*`, and then emacs freezes up and I cannot switch buffers any more.

**Comment by RalfJung** [comment]
@hendriktews 
> I cannot reproduce

Looks like you did not set the one setting which I identified as problematic above, `coq-compile-before-require`.

> Please provide more detailed instructions

Here's the minimal `.emacs`:
```
;; Custom (managed) Settings
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(coq-compile-before-require t))

;; Coq
(load "~/src/coq/ProofGeneral/generic/proof-site.el")
```

With that file, the following reproduces the problem, using Coq 8.11.2 from opam:
```
git clone  https://github.com/mit-pdos/perennial
cd perennial/
git submodule update --init --recursive
make -j8 external/Goose/github_com/mit_pdos/perennial_examples/replicated_block.vo
emacs external/Goose/github_com/mit_pdos/perennial_examples/replicated_block.v
```
This still takes a while to build, and interestingly it's not *all* files that are broken. I'll walk up the dependency chain.

In emacs, I got to the bottom of the file and do C-c C-Return.

**Comment by RalfJung** [comment]
> When you open the *Messages* buffer in a second frame, do you see
any pattern of messages scrolling by?

I managed to see something in that buffer by disabling "3-window mode" in the Coq menu. A *lot* of things scroll by faster than I can read; then things freeze, it looks as follows:

![Screenshot_20200612_134003](https://user-images.githubusercontent.com/330628/84499224-6cef4e00-acb2-11ea-85dc-fbe8f5f704c8.png)

**Comment by RalfJung** [comment]
Here's something strange:
```
rm external/iris/_CoqProject # needed to make PG pick up the right _CoqProject file
emacs external/iris/theories/base_logic/lib/fancy_updates.v
```
This doesn't eat up all my RAM, but it tops out at 11% of my 16GB of RAM. When I open the same file in a separate clone of Iris, RAM usage is much lower. Looks like the further I go up the dependency chain, the smaller RAM usage becomes, but it's still way bigger than I would expect.

One thing that is peculiar about the perennial repository is that it's a *ton* of files that are all built in one makefile. The overall dependency graph is rather big. So maybe this is not PG going into an infinite loop, maybe this is just exposing that PG is wasting a lot of RAM on dependency tracking, and this repository is big enough to actually make that a problem?

**Comment by hendriktews** [comment]
Thanks for your patience, I can now reproduce with `program_logic/crash_weakestpre`. And thanks for spotting that I forgot about the crucial option. Your suggestion `replicated_block` does not work here, because make fails on it. I can reproduce with only asserting the first two lines of `crash_weakestpre`. For them PG processes around 400 dependencies only, until I interrupt it with C-g. I therefore don't believe the amount of dependencies is the problem.

**Comment by hendriktews** [comment]
It will take a while to debug this. What I can see now is that the problem immediately starts with processing the coqdep output of `external/iris/theories/base_logic/upred.v`. Is there anything special with that file?

Meanwhile I have another question: Does the makefile also generate coq project files? Because, when I process some imports in a fresh checkout, it immediately fails, because libraries cannot be found. I would guess now, that the load path is incomplete, maybe, because the project files are not there yet?

**Comment by hendriktews** [comment]
With some more debug output the problems start one file earlier at `external/stdpp/theorie/proof_irrel.v`, it has probably nothing to do with the files.

**Comment by RalfJung** [comment]
> What I can see now is that the problem immediately starts with processing the coqdep output of external/iris/theories/base_logic/upred.v. Is there anything special with that file?

Hm, not that I know of.

> Meanwhile I have another question: Does the makefile also generate coq project files?

Looks like the topmost `_CoqProject` is generated by the `Makefile`, yes. Cc @tchajed

**Comment by hendriktews** [comment]
> Looks like the topmost _CoqProject is generated by the Makefile, yes. Cc @tchajed

OK, thanks. After `make _CoqProject` auto compilation works. But
this does not give any new insights.

**Comment by tchajed** [comment]
We do have a weird setup where the dependencies are hooked into a single tree and then we have a custom Makefile that just uses coqdep and has its own compilation rule. One issue this does raise is that there are multiple _CoqProject files, and only the root one should be used (which is what the Makefile does).

@RalfJung if your hypothesis is right that it's just the magnitude of the dependencies, you could try disabling compiling the large ones with this:
```el
(setq coq-compile-ignored-directories
      '("external/iris/.*" "external/stdpp/.*" "external/coqutil/.*"))
```

**Comment by hendriktews** [comment]
I found the problem: It's the naive way of propagating ancestors upwards in the dependency tree. When I programmed this, I simply appended all ancestors of all dependees - this leads to an exponential duplication of ancestors in tightly coupled dependency graphs.

If you want a quick solution, replace
```coq
  (let ((ancestors (get job 'ancestors))
```
in function `coq-par-kickoff-coqc-dependants` in `coq-par-compile.el` with

```coq
  (let ((ancestors (and coq-lock-ancestors (get job 'ancestors)))
```
and disable `coq-lock-ancestors`.

I will replace the lists with a hash, I guess.
